### PR TITLE
Enhancement

### PR DESCRIPTION
**Pull request by atifjavaid267**
- Refactor views code to remove extra spaces and add missing spaces etc.
- Move JS code to seperate files.
- Refactor the flow of apps action.
- Refactor the code for security.
- Ability set for product to all users
- Refactor home page
- Refactor root page
- Refactor button_to to link_to
- Image backed validation (active storage)
- Remove unused files and code

### CODE REVIEW

File: app/models/notification.rb [hunk 1]
```diff
@@ -10,4 +10,15 @@ def self.new_notification(sender_id, receiver_id)
...
+
+  def self.update_notifications_and_get_messages(current_user, sender_id)
+    notification = find_by(receiver_id: current_user.id, sender_id:)
+    notification.update(count: 0, read: true) if notification
+
+    messages = Message.where(sender_id: [current_user.id, sender_id],
+                             receiver_id: [current_user.id,
+                                           sender_id])
+
+    [notification, messages]
```
Comment (hassanakbar4) [inline app/models/notification.rb:22]: - Do we need to return notifications here? 
- Messages should not be filtered here.

File: app/models/notification.rb [hunk 2]
```diff
@@ -10,4 +10,15 @@ def self.new_notification(sender_id, receiver_id)
     end
     notification.count
   end
+
+  def self.update_notifications_and_get_messages(current_user, sender_id)
+    notification = find_by(receiver_id: current_user.id, sender_id:)
+    notification.update(count: 0, read: true) if notification
```
Comment (hassanakbar4) [inline app/models/notification.rb:16]: Need to check if count > 0

File: app/models/order.rb [hunk 3]
```diff
@@ -14,8 +14,6 @@ class Order < ApplicationRecord
 
   enum status: { pending: 0, successful: 1, cancelled: 2 }
 
-  scope :status, ->(status_param) { where(status: status_param) }
-
   private
 
   def pickup_time_cannot_be_in_the_past
```
Comment (hassanakbar4) [inline app/models/order.rb:19]: https://github.com/atifjavaid267/Online-Marketplace-for-Selling-Scrape/pull/80/files#diff-26bbec4be8ce33f60fa824e2bd29e6902e57541818578634bea06bce180b620eR32

Should be moved to scope

File: app/models/notification.rb [hunk 4]
```diff
@@ -1,4 +1,4 @@
-
+# class Notification
 class Notification < ApplicationRecord
   def self.new_notification(sender_id, receiver_id)
```
Comment (hassanakbar4) [inline app/models/notification.rb:3]: method name is not clear

**Review by hassana4 — CHANGES_REQUESTED** [review]

File: app/controllers/messages_controller.rb [hunk 5]
```diff
@@ -6,56 +6,19 @@ class MessagesController < ApplicationController
...
-    )
+    @sender_id = current_user.buyer? ? @order.seller_id : @order.buyer_id
+    @notification = Notification.update_notifications(current_user, @sender_id)
+    @messages = Message.messages(@current_user, @sender_id)
   end
 
   def create
     @message.sender_id = current_user.id
     @message.receiver_id = current_user.seller? ? @order.buyer_id : @order.seller_id
     return unless @message.save
```
Comment (HassanAkbar) [inline app/controllers/messages_controller.rb:21]: Error handling

File: app/controllers/users/unlocks_controller.rb [hunk 6]
```diff
@@ -1,3 +0,0 @@
-class Users::UnlocksController < Devise::UnlocksController
-
-end
```
Comment (HassanAkbar) [inline app/controllers/users/unlocks_controller.rb:3]: Break

File: app/models/message.rb [hunk 7]
```diff
@@ -1,13 +1,23 @@
...
+    where(sender_id: [current_user, sender_id], receiver_id: [current_user, sender_id])
+  end
+
+  private
 
-  def self.unread_message_count(sender_id, receiver_id)
-    where(receiver_id:, sender_id:, read_at: nil).count
+  def broadcast
+    MessageBroadcaster.new(self, order).broadcast_notifications
+    MessageBroadcaster.new(self, order).broadcast_room_channel
```
[the remark this answers is not in the document]
Comment (progammer743) [reply app/models/message.rb:21]: Ideal Service
https://medium.com/selleo/essential-rubyonrails-patterns-part-1-service-objects-1af9f9573ca1

File: app/channels/message_channel.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+class MessageChannel < ApplicationCable::Channel
+  def subscribed
+    stream_from "message_channel_#{params[:room_id]}"
```
Comment (HassanAkbar) [inline app/channels/message_channel.rb:3]: Why RoomID ?

File: app/javascript/channels/notification_channel.js [hunk 9]
```diff
@@ -109,36 +14,26 @@ document.addEventListener("turbolinks:load", () => {
...
+    const senderName = notificationMessages[pairId].senderName;
+    const messageCount = messageCounts[pairId];
+    const orderID = notificationMessages[pairId].orderID;
+
+    const notificationMessage = document.createElement("div");
+    notificationMessage.textContent = `${senderName} sent you ${messageCount} messages`;
+    notificationMessage.classList.add("p-2", "text-white");
+    notificationMessage.id = `notification-${pairId}`;
+    notificationMessage.addEventListener("click", () => {
+      const messageURL = `/orders/${orderID}/messages/new`;
```
Comment (HassanAkbar) [inline app/javascript/channels/notification_channel.js:28]: Either use orderID or remove it

Comment (HassanAkbar) [reply app/javascript/channels/notification_channel.js:28]: Nested route

File: app/models/notification.rb [hunk 10]
```diff
@@ -1,13 +1,21 @@
...
+    Notification.find_or_create_by(sender_id:, receiver_id:)
+  end
+
+  def increment_total
+    update(total: total + 1)
+  end
+
+  def total_notifications
+    total
+  end
```
Comment (HassanAkbar) [inline app/models/notification.rb:14]: Method not needed ?

File: app/models/notification.rb [hunk 11]
```diff
@@ -1,13 +1,21 @@
...
+
+  def self.clear_notifications(current_user_id, buyer_id, seller_id)
+    sender_id = current_user_id == seller_id ? buyer_id : seller_id
+    notification = find_by(receiver_id: current_user_id, sender_id:)
```
Comment (HassanAkbar) [inline app/models/notification.rb:18]: 1. Method name?
2. We should make an association in current_user for notifications using `reciever_id` as foreign_key
3. Scope

**Pull request merged**